### PR TITLE
Add no-buffering file logging and wait for a debugger option.

### PR DESCRIPTION
**Pull request by vsrs**
Adds two command line flags: `--no-buffering` and `--wait-dbg`. 

Not  sure if someone else needs this, but personally I found both flags extremely useful trying to figure out why RA does not work with Visual Studio. Or better to say why Visual Studio does not work with RA.

### CODE REVIEW

File: crates/rust-analyzer/src/bin/logger.rs [hunk 1]
```diff
@@ -63,11 +64,20 @@ impl Log for Logger {
...
     fn flush(&self) {
-        if let Some(w) = &self.file {
-            let _ = w.lock().flush();
+        match &self.file {
+            Some(w) => {
+                let _ = w.lock().flush();
+            }
+            None => {
+                let _ = io::stderr().flush();
+            }
```
Comment (matklad) [inline crates/rust-analyzer/src/bin/logger.rs:80]: Ahhhhhh....

This explains why logs misbehave for the last n months.

Let's just flush after every message unconditionally.

Comment (vsrs) [reply crates/rust-analyzer/src/bin/logger.rs:80]: > Let's just flush after every message unconditionally.

Sorry, but looks like a bad idea for me. It's super slow actually, for stderr too.
I think most of the time buffering not bad at all. And it's easy to add the flag if you need to.

Comment (matklad) [reply crates/rust-analyzer/src/bin/logger.rs:80]: Hm, that's surprisng, I don't think it should be slow even with sync enabled, as most messages are rather long. Could you double-check that it's the flushing that is slow? 

But maybe I am missing smth windows-specific here...

File: crates/rust-analyzer/src/bin/args.rs [hunk 2]
```diff
@@ -47,11 +49,16 @@ FLAGS:
     -vv, --spammy
     -q,  --quiet      Set verbosity
 
-    --log-file <PATH> Log to the specified filed instead of stderr
+    --log-file <PATH> Log to the specified file instead of stderr
+    --no-buffering    Flush log records to the file immediately
```
Comment (matklad) [inline crates/rust-analyzer/src/bin/args.rs:53]:
```suggestion
    --no-log-buffering  Flush log records to the file immediately
```

**Review by matklad — COMMENTED** [review]
bors d+

File: crates/rust-analyzer/src/bin/main.rs [hunk 3]
```diff
@@ -21,14 +21,25 @@ static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
...
-    setup_logging(args.log_file)?;
+
+    #[cfg(debug_assertions)]
+    if args.wait_dbg || env::var("RA_WAIT_DBG").is_ok() {
+        #[allow(unused_mut)]
+        let mut d = 4;
+        while d == 4 {
+            d = 4;
+        }
+    }
```
Comment (bjorn3) [inline crates/rust-analyzer/src/bin/main.rs:40]: Can you give some kind of indication that flag/env var is ignored when debug assertions are disabled? Or just panic?

File: crates/rust-analyzer/src/bin/main.rs [hunk 4]
```diff
@@ -21,14 +21,25 @@ static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
...
     let args = args::Args::parse()?;
-    setup_logging(args.log_file)?;
+
+    #[cfg(debug_assertions)]
+    if args.wait_dbg || env::var("RA_WAIT_DBG").is_ok() {
+        #[allow(unused_mut)]
+        let mut d = 4;
+        while d == 4 {
+            d = 4;
+        }
```
Comment (bjorn3) [inline crates/rust-analyzer/src/bin/main.rs:39]: What about an `#[inline(never)]` function with an infinite loop (`loop {}`) from which you can forcefully return? That should also work with optimizations I think.

Comment (vsrs) [reply crates/rust-analyzer/src/bin/main.rs:39]: The code has no sense for release builds, so it's great if the optimizer can throw it away automatically :)

File: crates/rust-analyzer/src/bin/args.rs (hunk 2 above)
Comment (vsrs) [reply crates/rust-analyzer/src/bin/args.rs]: > But maybe I am missing smth windows-specific here...

Whoops, my bad. Totally forgot that I minimized the pipe buffer. No lags with stderr at all, so I made the stderr flush unconditional.

File: crates/rust-analyzer/src/bin/main.rs (hunk 3 above)
Comment (vsrs) [reply crates/rust-analyzer/src/bin/main.rs:40]: Ah, I got it. `--wait-dbg` is pretty niche and to my mind it's better not to confuse users. So it is intended that in release builds we just ignore the flag.

**Pull request merged**